File: src/schur-weyl/tableaux.py
from typing import Generator
from young_diagrams import removable_corners


def standard_young_tableaux(lam: list[int]) -> Generator[list[list[int]]]:
    """Generates all SYT of a given shape

    Args:
        lam (list[int]): The partition

    Yields:
        Generator[list[list[int]]]: a SYT

    Examples:
        >>> list(standard_young_tableaux([]))
        [[]]
        >>> sorted(standard_young_tableaux([2, 1]))
        [[[1, 2], [3]], [[1, 3], [2]]]
        >>> len(list(standard_young_tableaux([3, 2, 1])))
        16
    """
    if not lam:
        yield []
        return

    ret: list[list[int]] = [[-1] * li for li in lam]
    ret[0][0] = 1

    def syt_rec(small_lam: list[int], d: int) -> Generator[list[list[int]]]:
        if len(small_lam) == 1 and small_lam[0] == 1:
            yield [row.copy() for row in ret]
            return
        for i, j in removable_corners(small_lam):
            if j == 0:
                small_lam.pop(i)
                ret[i][j] = d
                yield from syt_rec(small_lam, d - 1)
                small_lam.append(1)
            else:
                small_lam[i] -= 1
                ret[i][j] = d
                yield from syt_rec(small_lam, d - 1)
                small_lam[i] += 1

    yield from syt_rec(lam, sum(lam))


def semi_standard_young_tableau(lam: list[int], d: int) -> Generator[list[list[int]]]:
    """Generate all SSYT of a given shape and alphabet

    Args:
        lam (list[int]): The partition
        d (int): the alphabet size

    Yields:
        Generator[list[list[int]]]: The SSYT

    Examples:
        >>> list(semi_standard_young_tableau([2, 1], 3))
        [[[1, 1], [2]], [[1, 2], [2]], [[1, 3], [2]], [[1, 1], [3]], [[1, 2], [3]], [[1, 3], [3]], [[2, 2], [3]], [[2, 3], [3]]]
    """
    if not lam:
        yield []
        return
    ret: list[list[int]] = [[-1] * li for li in lam]
    maxv: list[list[int]] = [[d + 1] * li for li in lam]

    def fill(i: int, j: int, minv: int) -> Generator[list[list[int]]]:
        if j >= lam[i]:
            if i == 0:
                yield [row.copy() for row in ret]
                return
            yield from fill(i - 1, 0, i)
            return

        for cell_val in range(minv, maxv[i][j]):
            ret[i][j] = cell_val
            if i != 0:
                maxv[i - 1][j] = cell_val
            yield from fill(i, j + 1, cell_val)

    yield from fill(len(lam) - 1, 0, len(lam))


def semi_standard_young_tableau_by_content(lam: list[int], content: list[int]) -> Generator[list[list[int]]]:
    """Generate all SSYT with given content

    Args:
        lam (list[int]): The partition
        content (list[int]): The content (Should satisfy sum(lam) == sum(content))

    Yields:
        Generator[list[list[int]]]: a SSYT

    Examples:
        >>> list(semi_standard_young_tableau_by_content([2, 1], [2, 1]))
        [[[1, 1], [2]]]
        >>> list(semi_standard_young_tableau_by_content([2, 1], [1, 1, 1]))
        [[[1, 3], [2]], [[1, 2], [3]]]
        >>> list(semi_standard_young_tableau_by_content([], []))
        [[]]
    """
    if not sum(lam) == sum(content):
        return
    if not lam:
        yield []
        return

    d: int = len(content)
    contentcp: list[int] = content.copy()
    ret: list[list[int]] = [[-1] * li for li in lam]
    maxv: list[list[int]] = [[d + 1] * li for li in lam]

    def fill(i: int, j: int, minv: int) -> Generator[list[list[int]]]:
        if j >= lam[i]:
            if i == 0:
                yield [row.copy() for row in ret]
                return
            yield from fill(i - 1, 0, i)
            return

        for cell_val in range(minv, maxv[i][j]):
            if contentcp[cell_val - 1] <= 0:
                continue
            ret[i][j] = cell_val
            if i != 0:
                maxv[i - 1][j] = cell_val
            contentcp[cell_val - 1] -= 1
            yield from fill(i, j + 1, cell_val)
            contentcp[cell_val - 1] += 1

    yield from fill(len(lam) - 1, 0, len(lam))


# TODO implement with better algorithm (horizontal strips)
def kostka(lam: list[int], mu: list[int]) -> int:
    """Compute the kotska number

    Args:
        lam (list[int]): The partition
        mu (list[int]): content

    Returns:
        int: $K_{\\lambda \\mu}$

    Examples:
        >>> kostka([2, 1], [1, 1, 1])
        2
        >>> from math import factorial
        >>> kostka([10,1,1,1,1,1,1,1], [1] * 17) == factorial(17) / (17 * factorial(9) * factorial(7))
        True
    """
    return len(list(semi_standard_young_tableau_by_content(lam, mu)))


def reverse_reading_word(tableau: list[list[int]]) -> list[int]:
    """The reverse reading word of a tableau: right-to-left within each row,
    bottom row to top row.

    Examples:
        >>> reading_word([[1, 1, 3], [2, 3], [4]])
        [4, 3, 2, 3, 1, 1]
        >>> reading_word([[1, 2], [3]])
        [3, 2, 1]
        >>> reading_word([])
        []
    """
    return [v for row in reversed(tableau) for v in reversed(row)]

File: docs/src/schur-weyl/tableaux.ts
/**
 * Ported from `src/schur_weyl/tableaux.py`.
 */

import type { Partition } from './youngDiagrams';
import { removableCorners } from './youngDiagrams';

/** A (semi-)standard Young tableau: rows of entries, ragged like the shape. */
export type Tableau = number[][];

function cloneTableau(t: Tableau): Tableau {
  return t.map((row) => [...row]);
}

/**
 * Generate all standard Young tableaux of a given shape.
 *
 * @example
 * [...standardYoungTableaux([])] // [[]]
 * [...standardYoungTableaux([2, 1])] // [[[1,2],[3]], [[1,3],[2]]]
 * [...standardYoungTableaux([3, 2, 1])].length // 16
 */
export function* standardYoungTableaux(lam: Partition): Generator<Tableau, void, void> {
  if (lam.length === 0) {
    yield [];
    return;
  }

  const ret: number[][] = lam.map((li) => new Array<number>(li).fill(-1));
  ret[0][0] = 1;

  function* sytRec(smallLam: number[], d: number): Generator<Tableau, void, void> {
    if (smallLam.length === 1 && smallLam[0] === 1) {
      yield cloneTableau(ret);
      return;
    }
    for (const [i, j] of removableCorners(smallLam)) {
      if (j === 0) {
        smallLam.splice(i, 1);
        ret[i][j] = d;
        yield* sytRec(smallLam, d - 1);
        smallLam.splice(i, 0, 1);
      } else {
        smallLam[i] -= 1;
        ret[i][j] = d;
        yield* sytRec(smallLam, d - 1);
        smallLam[i] += 1;
      }
    }
  }

  yield* sytRec([...lam], lam.reduce((a, b) => a + b, 0));
}

/**
 * Generate all semi-standard Young tableaux of a given shape and alphabet.
 *
 * @param d - the alphabet size
 *
 * @example
 * [...semiStandardYoungTableau([2, 1], 3)]
 * // [[[1,1],[2]], [[1,2],[2]], [[1,3],[2]], [[1,1],[3]], [[1,2],[3]],
 * //  [[1,3],[3]], [[2,2],[3]], [[2,3],[3]]]
 */
export function* semiStandardYoungTableau(lam: Partition, d: number): Generator<Tableau, void, void> {
  if (lam.length === 0) {
    yield [];
    return;
  }
  const ret: number[][] = lam.map((li) => new Array<number>(li).fill(-1));
  const maxv: number[][] = lam.map((li) => new Array<number>(li).fill(d + 1));

  function* fill(i: number, j: number, minv: number): Generator<Tableau, void, void> {
    if (j >= lam[i]) {
      if (i === 0) {
        yield cloneTableau(ret);
        return;
      }
      yield* fill(i - 1, 0, i);
      return;
    }
    for (let cellVal = minv; cellVal < maxv[i][j]; cellVal++) {
      ret[i][j] = cellVal;
      if (i !== 0) maxv[i - 1][j] = cellVal;
      yield* fill(i, j + 1, cellVal);
    }
  }

  yield* fill(lam.length - 1, 0, lam.length);
}

/**
 * Generate all semi-standard Young tableaux with given content.
 *
 * @param content - should satisfy sum(lam) === sum(content)
 *
 * @example
 * [...semiStandardYoungTableauByContent([2, 1], [2, 1])] // [[[1,1],[2]]]
 * [...semiStandardYoungTableauByContent([2, 1], [1, 1, 1])]
 * // [[[1,3],[2]], [[1,2],[3]]]
 * [...semiStandardYoungTableauByContent([], [])] // [[]]
 */
export function* semiStandardYoungTableauByContent(
  lam: Partition,
  content: number[],
): Generator<Tableau, void, void> {
  const lamSum = lam.reduce((a, b) => a + b, 0);
  const contentSum = content.reduce((a, b) => a + b, 0);
  if (lamSum !== contentSum) return;
  if (lam.length === 0) {
    yield [];
    return;
  }

  const d = content.length;
  const contentCp = [...content];
  const ret: number[][] = lam.map((li) => new Array<number>(li).fill(-1));
  const maxv: number[][] = lam.map((li) => new Array<number>(li).fill(d + 1));

  function* fill(i: number, j: number, minv: number): Generator<Tableau, void, void> {
    if (j >= lam[i]) {
      if (i === 0) {
        yield cloneTableau(ret);
        return;
      }
      yield* fill(i - 1, 0, i);
      return;
    }
    for (let cellVal = minv; cellVal < maxv[i][j]; cellVal++) {
      if (contentCp[cellVal - 1] <= 0) continue;
      ret[i][j] = cellVal;
      if (i !== 0) maxv[i - 1][j] = cellVal;
      contentCp[cellVal - 1] -= 1;
      yield* fill(i, j + 1, cellVal);
      contentCp[cellVal - 1] += 1;
    }
  }

  yield* fill(lam.length - 1, 0, lam.length);
}

// TODO implement with a better algorithm (horizontal strips)
/**
 * Compute the Kostka number K_{lam, mu}.
 *
 * @example
 * kostka([2, 1], [1, 1, 1]) // 2
 */
export function kostka(lam: Partition, mu: Partition): number {
  return [...semiStandardYoungTableauByContent(lam, mu)].length;
}

/**
 * The reverse reading word of a tableau: right-to-left within each row,
 * bottom row to top row.
 *
 * @example
 * reverseReadingWord([[1,1,3],[2,3],[4]]) // [4,3,2,3,1,1]
 * reverseReadingWord([[1,2],[3]]) // [3,2,1]
 * reverseReadingWord([]) // []
 */
export function reverseReadingWord(tableau: Tableau): number[] {
  const result: number[] = [];
  for (let r = tableau.length - 1; r >= 0; r--) {
    const row = tableau[r];
    for (let c = row.length - 1; c >= 0; c--) result.push(row[c]);
  }
  return result;
}
